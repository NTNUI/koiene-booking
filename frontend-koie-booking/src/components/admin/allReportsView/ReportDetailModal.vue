<template>
  <v-dialog :value="showModal" persistent width="500" @click:outside="closeModal">
    <v-card>
      <v-card-title class="headline">
        {{ title }}
      </v-card-title>
      <v-card-text>
        <slot name="modalContext" />
      </v-card-text>
      <v-card-actions>
        <v-spacer></v-spacer>
        <v-btn color="green darken-1" text @click="closeModal">
          Lukk
        </v-btn>
      </v-card-actions>
    </v-card>
  </v-dialog>
</template>

<script>
import Vue from 'vue';

export default Vue.extend({
  name: 'ReportDetailModal',
  props: {
    showModal: {
      type: Boolean,
      default: true,
    },
    closeModal: {
      type: Function,
      default() {
        return null;
      },
    },
    title: {
      type: String,
      default: '',
    },
  },
});
</script>

<style scoped></style>
