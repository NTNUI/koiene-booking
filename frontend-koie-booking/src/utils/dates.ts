import dayjs from 'dayjs';
require('dayjs/locale/nb');
dayjs.locale('nb');

export function getDateString(baseDate?: string, addDays?: number): string {
  let resultString: string;
  if (baseDate && addDays) {
    resultString = new Date(new Date(baseDate).getTime() + addDays * 24 * 60 * 60 * 1000).toISOString().substr(0, 10);
  } else if (baseDate) {
    resultString = new Date(baseDate).toISOString().substr(0, 10);
  } else if (addDays) {
    resultString = new Date(new Date().getTime() + addDays * 24 * 60 * 60 * 1000).toISOString().substr(0, 10);
  } else {
    resultString = new Date().toISOString().substr(0, 10);
  }
  return resultString;
}

export function getDate(baseDate?: string, addDays?: number): Date {
  let resultDate: Date;
  if (baseDate && addDays) {
    resultDate = new Date(new Date(baseDate).getTime() + addDays * 24 * 60 * 60 * 1000);
  } else if (baseDate) {
    resultDate = new Date(baseDate);
  } else if (addDays) {
    resultDate = new Date(new Date().getTime() + addDays * 24 * 60 * 60 * 1000);
  } else {
    resultDate = new Date();
  }
  return resultDate;
}

export function addToDate(startDate: string, howMany: number, timeUnit: dayjs.OpUnitType): string {
  const date = dayjs(startDate + 'T00:00:00.000Z');
  return date.add(howMany, timeUnit).format('YYYY-MM-DD');
}

export function formatDate(dateISO: string, formatString: string): string {
  const date = dayjs(dateISO + 'T00:00:00.000Z');
  return date.format(formatString);
}
