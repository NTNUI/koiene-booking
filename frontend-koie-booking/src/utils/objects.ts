export const clone = (object: any) => Object.assign({}, object);

export const renameKey = (object: any, key: string, newKey: string): any => {
  const value = object[key];
  delete object[key];
  object[newKey] = value;
  return object;
};

export const sortObjectByKey = (unordered: any) => {
  let ordered: any;
  ordered = {};
  Object.keys(unordered)
    .sort()
    .forEach(function(key) {
      ordered[key] = unordered[key];
    });
  return ordered;
};
