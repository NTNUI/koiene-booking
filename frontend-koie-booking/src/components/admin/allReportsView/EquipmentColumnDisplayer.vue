<template>
  <div>
    <v-row justify="center">
      <v-chip class="mx-1" :color="$scssVars.globalColorGreenStrong" @click="showDialog = true">
        {{ ok.length }}
      </v-chip>
      <v-chip class="mx-1" :color="$scssVars.globalColorGrey" @click="showDialog = true"> {{ notSure.length }} </v-chip>
      <v-chip class="mx-1" :color="$scssVars.globalColorRedStrong" @click="showDialog = true">
        {{ notOk.length }}
      </v-chip>
    </v-row>
    <ReportDetailModal title="Utstyr" :show-modal="showDialog" :close-modal="closeDialog">
      <template v-slot:modalContext>
        <v-row justify="center">
          <v-col cols="4" :style="'color:' + $scssVars.globalColorGreenStrong">
            <h3>OK</h3>
            <ul>
              <li v-for="(okItem, index) in ok" :key="'ok' + index">{{ $t('report.equipment.' + okItem) }}</li>
            </ul>
          </v-col>
          <v-col cols="4" :style="'color:' + $scssVars.globalColorGrey">
            <h3>Ikke sikker</h3>
            <ul>
              <li v-for="(notSureItem, index) in notSure" :key="'notSure' + index">
                {{ $t('report.equipment.' + notSureItem) }}
              </li>
            </ul>
          </v-col>
          <v-col cols="4" :style="'color:' + $scssVars.globalColorRedStrong">
            <h3>Ikke OK</h3>
            <ul>
              <li v-for="(notOkItem, index) in notOk" :key="'notOkItem' + index">
                {{ $t('report.equipment.' + notOkItem) }}
              </li>
            </ul>
          </v-col>
        </v-row>
      </template>
    </ReportDetailModal>
  </div>
</template>

<script lang="ts">
import Vue from 'vue';
import ReportDetailModal from '@/components/admin/allReportsView/ReportDetailModal.vue';

export default Vue.extend({
  name: 'EquipmentColumnDisplayer',
  components: { ReportDetailModal },
  props: {
    notOk: {
      type: Array,
      default() {
        return [];
      },
    },
    notSure: {
      type: Array,
      default() {
        return [];
      },
    },
    ok: {
      type: Array,
      default() {
        return [];
      },
    },
  },
  data() {
    return {
      showDialog: false,
    };
  },
  methods: {
    closeDialog() {
      this.showDialog = false;
    },
  },
});
</script>

<style scoped></style>
