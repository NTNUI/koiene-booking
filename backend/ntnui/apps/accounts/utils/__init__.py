from accounts.utils.mail_utils import send_gsuite_confirmation_mail
