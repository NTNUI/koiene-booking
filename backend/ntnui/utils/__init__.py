from ntnui.utils.pagination import PageSize7, PageSize20, PageSize100
