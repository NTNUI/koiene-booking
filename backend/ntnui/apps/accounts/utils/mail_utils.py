from django.core.mail import send_mail
from django.template.loader import render_to_string

from ntnui.enums import MembershipType


def new_membership(membership):
    """ Sends confirmation email to users and guests who signs up for events and sub-events. """

    # Mail header.
    sender = "noreply@mg.ntnui.no"
    receiver = [membership.member.email]
    subject = "Welcome"

    context = {"user": membership.member, "group": membership.group}

    # Mail body.
    plain_message = render_to_string("emails/new_membership.txt", context)
    html_message = render_to_string("emails/new_membership.html", context)

    send_mail(subject, plain_message, sender, receiver, html_message=html_message)


def send_gsuite_confirmation_mail(user, contact_email, password):
    # Mail header.
    sender = "noreply@mg.ntnui.no"
    receiver = [user.email]
    subject = "Opprettelse av ny NTNUI-bruker"

    context = {
        "first_name": user.first_name,
        "email": user.email,
        "contact_email": contact_email,
        "password": password,
    }

    # Mail body.
    plain_message = render_to_string("emails/new_gsuite_account.txt", context)
    html_message = render_to_string("emails/new_gsuite_account.html", context)

    send_mail(subject, plain_message, sender, receiver, html_message=html_message)


def send_welcome_mail_to_leader(user, group):
    """ Send a mail to a new leader """
    sender = "noreply@mg.ntnui.no"
    receiver = [user.email]
    subject = f"Velkommen som ny leder i {group}"

    context = {
        "first_name": user.first_name,
        "email": user.get_contact_email(),
        "group": group.name,
    }

    plain_message = render_to_string(
        template_name="emails/new_leader_membership.txt", context=context
    )
    html_message = render_to_string(
        template_name="emails/new_leader_membership.html", context=context
    )

    send_mail(
        subject=subject,
        message=plain_message,
        from_email=sender,
        recipient_list=receiver,
        html_message=html_message,
    )


def send_welcome_mail_to_cashier(user, group):
    """Send a mail to the new cashier"""
    sender = "noreply@mg.ntnui.no"
    receiver = [user.email]
    subject = f"Velkommen som ny kasserer i {group}"

    context = {
        "first_name": user.first_name,
        "email": user.get_contact_email(),
        "group": group.name,
    }

    plain_message = render_to_string(
        template_name="emails/new_cashier_membership.txt", context=context
    )
    html_message = render_to_string(
        template_name="emails/new_cashier_membership.html", context=context
    )

    send_mail(
        subject=subject,
        message=plain_message,
        from_email=sender,
        recipient_list=receiver,
        html_message=html_message,
    )


def send_welcome_mail_to_board_member(user, group):
    """Send a mail to the new cashier"""
    sender = "noreply@mg.ntnui.no"
    receiver = [user.email]
    subject = f"Velkommen som ny tillitsvalgt i {group}"

    context = {
        "first_name": user.first_name,
        "email": user.get_contact_email(),
        "group": group.name,
    }

    plain_message = render_to_string(
        template_name="emails/new_board_membership.txt", context=context
    )
    html_message = render_to_string(
        template_name="emails/new_board_membership.html", context=context
    )

    send_mail(
        subject=subject,
        message=plain_message,
        from_email=sender,
        recipient_list=receiver,
        html_message=html_message,
    )


def send_welcome_mail(membership):
    """ Delegate welcome mail to different board members """
    send_welcome_mail_to_board_member(user=membership.member, group=membership.group)
    if membership.type == MembershipType.leader:
        send_welcome_mail_to_leader(user=membership.member, group=membership.group)

    if membership.type == MembershipType.cashier:
        send_welcome_mail_to_cashier(user=membership.member, group=membership.group)
