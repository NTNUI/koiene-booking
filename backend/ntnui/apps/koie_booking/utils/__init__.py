from koie_booking.utils.mail_utils import send_confirmation_mail, send_koie_information_mail
from koie_booking.utils.date_utils import get_daterange
