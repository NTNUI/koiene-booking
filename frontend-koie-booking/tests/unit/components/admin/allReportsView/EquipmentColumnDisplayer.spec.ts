import Vue from 'vue';
import Vuetify from 'vuetify';
import { createShallowWrapper } from '../../../utils';
import { Wrapper } from '@vue/test-utils';

import EquipmentColumnDisplayer from '@/components/admin/allReportsView/EquipmentColumnDisplayer.vue';
import ReportDetailModal from '@/components/admin/allReportsView/ReportDetailModal.vue';

Vue.use(Vuetify);

describe('Component EquipmentColumnDisplayer', () => {
  let wrapper: Wrapper<any>;

  beforeEach(() => {
    const wrapperOptions = {
      propsData: {
        notOK: ['axe', 'hammer'],
        notSure: ['saw', 'saw_blade', 'spade'],
        ok: ['detergent'],
      },
      data() {
        return { showDialog: true };
      },
    };
    wrapper = createShallowWrapper(EquipmentColumnDisplayer, wrapperOptions);
  });

  it('Matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Shows dialog when notSureChip is clicked', () => {
    const notSureChip = wrapper.find({ ref: 'equipmentColumnDisplayerNotSureChip' });
    notSureChip.trigger('click');

    expect(wrapper.vm.showDialog).toBeTruthy();
  });

  it('Closes dialog', () => {
    wrapper.vm.closeDialog();

    expect(wrapper.vm.showDialog).toBeFalsy();
  });

  it('Shows modal when showDialog is true', () => {
    expect(wrapper.contains(ReportDetailModal)).toBeTruthy();
  });

  it('Shows default values when no inputs are given', () => {
    wrapper = createShallowWrapper(EquipmentColumnDisplayer);
    expect(wrapper.vm.notOk).toEqual([]);
    expect(wrapper.vm.notSure).toEqual([]);
    expect(wrapper.vm.ok).toEqual([]);
    expect(wrapper.vm.showDialog).toBeFalsy();
  });
});
