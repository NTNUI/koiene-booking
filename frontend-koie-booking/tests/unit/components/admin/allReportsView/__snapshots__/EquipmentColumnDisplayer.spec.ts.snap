// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component EquipmentColumnDisplayer Matches snapshot 1`] = `
<div notok="axe,hammer">
  <v-row-stub tag="div" justify="center">
    <v-chip-stub activeclass="" tag="span" active="true" closeicon="$delete" filtericon="$complete" class="mx-1">
      1
    </v-chip-stub>
    <v-chip-stub activeclass="" tag="span" active="true" closeicon="$delete" filtericon="$complete" class="mx-1">
      3
    </v-chip-stub>
    <v-chip-stub activeclass="" tag="span" active="true" closeicon="$delete" filtericon="$complete" class="mx-1">
      0
    </v-chip-stub>
  </v-row-stub>
  <reportdetailmodal-stub showmodal="true" closemodal="function () { [native code] }" title="Utstyr"></reportdetailmodal-stub>
</div>
`;
